### PR TITLE
Fault tolerant distributed training, just work version, with etcd

### PR DESCRIPTION
**Pull request by typhoonzero**
This is a "just work" version, supporting distributed training with etcd and master process.

### CODE REVIEW

File: go/master/client.go [hunk 1]
```diff
@@ -38,7 +39,8 @@ func (c *Client) getRecords() {
 		if err != nil {
 			// TODO(helin): wait before move on with next
```
Comment (helinwang) [inline go/master/client.go]: Could you remove the TODO, since it's completed?

File: go/pserver/client/c/test/test_train.py [hunk 2]
```diff
@@ -1,5 +1,23 @@
...
 import paddle.v2.dataset.uci_housing as uci_housing
+import paddle.v2.master as master
+import os
+import cPickle as pickle
+
+etcd_ip = os.getenv("MASTER_IP", "127.0.0.1")
+etcd_endpoint = "http://" + etcd_ip + ":2379"
+
+
+def cloud_reader():
```
Comment (helinwang) [inline go/pserver/client/c/test/test_train.py]: We can put this into `paddle.reader.recordio` in the future, so `paddle.reader.recordio` can support both local and cloud reader. (not necessary in this PR).

File: python/paddle/v2/trainer.py [hunk 3]
```diff
@@ -51,7 +51,8 @@ def __init__(self,
                  update_equation,
                  extra_layers=None,
                  is_local=True,
-                 pserver_spec=None):
+                 pserver_spec=None,
+                 use_etcd=True):
```
Comment (helinwang) [inline python/paddle/v2/trainer.py]: Maybe we can make the default requires less dependency, by defaulting `use_etcd` to False.

File: go/master/client.go (hunk 1 above)
Comment (typhoonzero) [inline go/master/client.go]: Done.

File: go/pserver/client/c/test/test_train.py (hunk 2 above)
Comment (typhoonzero) [inline go/pserver/client/c/test/test_train.py]: Yes! Will do this in next PR.

File: python/paddle/v2/trainer.py (hunk 3 above)
Comment (typhoonzero) [inline python/paddle/v2/trainer.py]: `use_etcd` has default value to be `True` currently

File: go/pserver/client/c/test/test_train.py [hunk 4]
```diff
@@ -1,5 +1,23 @@
 import paddle.v2 as paddle
 import paddle.v2.dataset.uci_housing as uci_housing
+import paddle.v2.master as master
+import os
+import cPickle as pickle
+
+etcd_ip = os.getenv("MASTER_IP", "127.0.0.1")
```
Comment (helinwang) [inline go/pserver/client/c/test/test_train.py]: 不好意思，这里上次没看见，是不是应该把“MASTER_IP”改成"ETCD_IP"? master貌似跟etcd没有关系。

File: paddle/trainer/NewRemoteParameterUpdater.cpp [hunk 5]
```diff
@@ -28,6 +28,17 @@ NewRemoteParameterUpdater::NewRemoteParameterUpdater(
...
 
+NewRemoteParameterUpdater::NewRemoteParameterUpdater(
+    const OptimizationConfig &config,
+    const std::string pserverSpec,
+    const bool useEtcd)
+    : trainerConfig_(config),
+      parameterClient_(-1),
+      newParameters_(nullptr),
+      newGradients_(nullptr),
+      pserverSpec_(pserverSpec),
```
Comment (helinwang) [inline paddle/trainer/NewRemoteParameterUpdater.cpp]: 这里有时候是pserver的addresses，有时候是etcd的address，是不是该改成remoteAddr之类的。

File: go/pserver/client/c/test/test_train.py (hunk 4 above)
Comment (typhoonzero) [inline go/pserver/client/c/test/test_train.py]: 目前paddlecloud启动job时，etcd是启动在master pod里的。后续考虑多个job使用同一个etcd的情况下修改下。

**Pull request merged**